File: html-parse/Source Code/HtmlToListParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using HtmlAgilityPack;

namespace KariyerDotNet
{
    class HtmlToListParser
    {

        private static List<string> jobDescDelimiters = new List<string>()
        {
            "İŞ TANIMI",
            "İŞ TANITIMI",
            "İŞ GEREKSİNİMLERİ",
            "GENEL İŞ TANIMI",
            "İŞ ÖZELLİKLERİ",
            "İŞ AÇIKLAMASI"
        };

        private static List<string> employeeDescDelimiters = new List<string>()
        {
            "GENEL TANITIM",
            "GENEL TANIM",
            "GENEL ÖZELLİKLER",
            "GENEL NİTELİKLER",
            "GENEL YETENEKLER",
            "İSTENEN YETENEK VE UZMANLIKLAR"
        };

        private static int FAIL_COUNTER = 0;

        /**
         * This function will decide to parse the string as;
         *      Job and Employee texts differently
         *      HTML Lists or Break Lines 
         * 
         */
        public static List<string> parse(string metin)
        {
            string jobDescDelimiter = String.Empty;
            string employeeDescDelimiter = String.Empty;
            foreach (string delimiter in jobDescDelimiters)
            {
                
                if (metin.ToUpper().Contains(delimiter))
                {
                    jobDescDelimiter = delimiter;
                    break;
                }
            }
            string jobText = String.Empty;
            string employeeText = String.Empty;
            bool hasDoubleText = true;
            if (!String.IsNullOrEmpty(jobDescDelimiter))
            {
                string[] temp = Regex.Split(metin, jobDescDelimiter, RegexOptions.IgnoreCase);
                if (temp[0].Trim() != String.Empty)
                {
                    // İş tanıtımı bölümü genel nitelikler bölümünden sonra geliyor.
                    jobText = temp[0];
                    employeeText = temp[1];
                }
                else
                {
                    // İş tanıtımı bölümü genel nitelikler bölümünden önce geliyor
                    // Genel nitelikler göz önününde bulundurularak ayrıştırılma yapılması gerek
                    foreach (string delimiter in employeeDescDelimiters)
                    {
                        if (metin.ToUpper().Contains(delimiter))
                        {
                            employeeDescDelimiter = delimiter;
                            break;
                        }
                    }
                    if (!String.IsNullOrEmpty(employeeDescDelimiter))
                    {
                        string[] empTemp = Regex.Split(metin, employeeDescDelimiter, RegexOptions.IgnoreCase);
                        if (empTemp[0].Trim() != String.Empty)
                        {
                            jobText = empTemp[1];
                            employeeText = empTemp[0];
                        }
                        else
                        {
                            // This case shouldn't be happening that much;
                            hasDoubleText = false;
                            FAIL_COUNTER++;
                        }
                    }
                    else
                    {
                        hasDoubleText = false;
                    }
                }
            }
            else
            {
                hasDoubleText = false;
            }
            List<string> result = null;
            if (!hasDoubleText)
            {
                result = textParse(metin);
            }
            else
            {
                result = textParse(jobText);
                //Console.WriteLine(jobText);
                //result.AddRange(textParse(employeeText));
            }
            return result;
        }


        private static List<string> textParse(string text)
        {
            List<string> items = new List<string>();
            bool containsLists = text.Contains("<li>");
            bool containsBreakLines = text.Contains("<br>");

            if (containsLists)
            {
                loadListTags(items, text);
            }

            if (containsBreakLines)
            {
                string[] texts = Regex.Split(text, "<br>", RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
                HashSet<int> listIndexes = new HashSet<int>();
                // Try to parse by list items, in break lines splitted texts.
                // If there is none, do nothing. Continue with break line listing.
                try
                {
                    for (int i = 0; i < texts.Length; i++)
                    {
                        string _text = texts[i];
                        // If that portion doesn't contain lists, include it with tags removed.
                        if (!_text.Contains("<li>") && !_text.Contains("</li>"))
                        {
                            //This regex replaces all tags with empty string.
                            _text = changeHTMLChars(_text).Trim();
                            _text = Regex.Replace(_text, "<.*?>", String.Empty);
                            // Dont include any text that is longer than 170 characters AND is ALL upper case
                            if(_text.Length <= 170 && !_text.ToUpper().Equals(_text))
                            {
                                items.Add(_text);
                            }
                        }
                    }
                }
                catch (NullReferenceException) { /* DO NOTHING. THIS CATCH SHOULDN'T BE WORKING AT ALL */ }
            }
            return items;
        }

        private static void loadListTags(List<string> list, string text)
        {
            var htmlDoc = new HtmlDocument();
            htmlDoc.LoadHtml(text);
            // Select all <li> tags
            var listItemNodes = htmlDoc.DocumentNode.SelectNodes("//li");

            foreach (var listItemNode in listItemNodes)
            {
                String txt = changeHTMLChars(listItemNode.InnerText);
                // Dont include any text that is longer than 240 characters AND is ALL upper case
                if (txt.Length < 240 && !txt.ToUpper().Equals(txt))
                {
                    list.Add(txt);
                }
            }
        }
        /**
         * Change such HTML characters such as
         *      &lt; &gt; &nbsp;
         *      to actual characters.
         */ 
        private static string changeHTMLChars(string text)
        {
            text = text.Replace("&nbsp;", " ").Replace("&lt;", "<").Replace("&gt;", ">").Replace("·", String.Empty);
            return text;
        }
    }
}

File: src/main/java/edu/deu/seniorproject/parser/HtmlToListParser.java
package edu.deu.seniorproject.parser;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.regex.Pattern;

public class HtmlToListParser {

    private static final List<String> jobDescDelimiters = Arrays.asList(
            "İŞ TANIMI",
            "İŞ TANITIMI",
            "İŞ GEREKSİNİMLERİ",
            "GENEL İŞ TANIMI",
            "İŞ ÖZELLİKLERİ",
            "İŞ AÇIKLAMASI");

    private static final List<String> employeeDescDelimiters = Arrays.asList(
            "GENEL TANITIM",
            "GENEL TANIM",
            "GENEL ÖZELLİKLER",
            "GENEL NİTELİKLER",
            "GENEL YETENEKLER",
            "İSTENEN YETENEK VE UZMANLIKLAR");


    private int failCount;

    public HtmlToListParser() {
        failCount = 0;
    }

    public List<String> parse(String text) {
        String jobDescDelimiter = "";
        String employeeDescDelimiter = "";
        String jobText = "";
        String employeeText = "";

        boolean hasDoubleText = true;

        for (String delimiter : jobDescDelimiters) {
            if (text.toUpperCase().contains(delimiter)) {
                jobDescDelimiter = delimiter;
                break;
            }
        }

        if (!jobDescDelimiter.isEmpty()) {
            //Regex for splitting  case insensitive
            Pattern pattern = Pattern.compile(jobDescDelimiter, Pattern.CASE_INSENSITIVE | Pattern.UNICODE_CASE);
            String[] temp = pattern.split(text);
            if (!temp[0].trim().isEmpty()) {
                jobText = temp[0];
                employeeText = temp[1];
            } else {
                // Job description comes after wanted qualifications
                // Splitting must be done in terms of qualifications, not job description
                for (String delimiter : employeeDescDelimiters) {
                    if (text.toUpperCase().contains(delimiter)) {
                        employeeDescDelimiter = delimiter;
                        break;
                    }
                }
                if (!employeeDescDelimiter.isEmpty()) {
                    pattern = Pattern.compile(employeeDescDelimiter, Pattern.CASE_INSENSITIVE | Pattern.UNICODE_CASE);
                    String[] empTemp = pattern.split(text);
                    if (!empTemp[0].trim().isEmpty()) {
                        jobText = empTemp[1];
                        employeeText = empTemp[0];
                    } else {
                        // This case is extremely rare for the give job ads, and should not happen that much
                        // This case is ignored and contents of it isn't parsed.
                        hasDoubleText = false;
                        failCount++;
                    }
                } else {
                    hasDoubleText = false;
                }
            }
        } else {
            hasDoubleText = false;
        }

        List<String> result;
        if (!hasDoubleText) {
            // Parse the whole text
            result = textParse(text);
        } else {
            // Only parse the job information / description
            result = textParse(jobText);
        }
        return result;
    }

    private List<String> textParse(String text) {

        List<String> items = new ArrayList<>();
        boolean containsLists = text.contains("<li>");
        boolean containsBreakLines = text.contains("<br>");

        if (containsLists) {
            loadListTags(items, text);
        }

        if (containsBreakLines) {
            Pattern pattern = Pattern.compile("<br>", Pattern.CASE_INSENSITIVE | Pattern.UNICODE_CASE);
            String[] words = pattern.split(text);
            try {
                for (String word : words) {
                    // Double check for HTML listings
                    if (!word.contains("<li>") && !word.contains("</li>")) {
                        // This regex replaces all tags with an empty string
                        word = changeHTMLChars(word);
                        word = word.replaceAll("<.*?>", "");
                        // Don't include any text that is longer than 170 characters AND is ALL upper case
                        if (word.length() <= 170 && !word.toUpperCase().equals(word)) {
                            items.add(word);
                        }
                    }
                }

            } catch (Exception ignore) { /* Ignore this exception */ }
        }
        return items;
    }


    private void loadListTags(List<String> list, String text) {
        Document doc = Jsoup.parse(text);
        // This gets all list elements
        Elements elements = doc.body().select("li");
        for (Element element : elements) {
            String innerText = changeHTMLChars(element.text());
            if (innerText.length() < 240 && !innerText.toUpperCase().equals(innerText)) {
                list.add(innerText);
            }
        }
    }

    private String changeHTMLChars(String text) {
        text = text.replace("&nbsp;", " ").replace("&lt;", "<").
                replace("&gt;", ">").replace("·", "");
        return text.trim();
    }

}
